driver/asana_driver.py

```diff
@@ -0,0 +1,185 @@
+            identifier that will allow the subsequent run to pick up indexing
+            files where it last left off.
+
+        :returns `RetrieveMetadataResult` with the appropriate values populated.
+        """
+        
+        if milestone is None: milestone = { 'step':'stage1' } # TODO: verify use of milestone
+        
+        docs = []
+        
+        try:
+            for t in self.iter_tasks():
+                task = self.client.tasks.find_by_id(task=t['id'])
+                
+                # build document meta for task
+                doc = {
+                    'id' : task['id'], # TODO: verify id to be used
+                    'external_id' : task['id'], # TODO: verify id to be used
+                    'dirty' : True,
+                    'title' : t['name'],
+                    'url': 'https://app.asana.com/0/{project_id}/{task_id}'.format(project_id=t['project']['id'], task_id=t['id']),
+                    'created': parse_date( task['created_at'] ),
+                    'edited': parse_date( task['modified_at'] ),
+                    'tag': map(lambda tag: tag['name'], task['tags']),
+                    'as_assignee': [ task['assignee']['name'] , task['assignee']['id'] ] if task['assignee'] else None,
+                    'as_completed': task['completed'],
+                    'as_completed_date': parse_date( task['completed_at'] ),
+                    'as_hearted': task['hearted'],
+                }
+                docs.append(doc) ; pp.pprint(doc)
+                
+                # get attachments for task
+                attachments = self.client.attachments.find_by_task(task=t['id'])
+                
+                # loop through attachments, and add to document pool
+                for a in attachments:
+                    attachment = self.client.attachments.find_by_id(attachment=a['id'])
+                    
+                     # build document meta for attachment
+                    doc = {
+                        'id' : attachment['id'], # TODO: verify id to be used
+                        'external_id' : attachment['id'], # TODO: verify id to be used
+                        'subtype' : 'attachment',
+                        'title' : attachment['name'],
+                        'url' : attachment['download_url'],
+                        'created': parse_date( attachment['created_at'] ),
+                        'path': [ attachment['parent']['name'] ],
+                        'container_url' : attachment['view_url'],
+                    }
+                    docs.append(doc) ; pp.pprint(doc)
+                    pass
+                pass
+        except asana.error.RateLimitEnforcedError, e:
+            raise RateLimitError( e.message )
+            
+        # build result object
+        result = RetrieveMetadataResult(
+            milestone=milestone,
+            retrieve_metadata_done=True,
+            docs=docs
+        )
+        
+        return result
```

Comment:
It appears as if repeated calls to the function will duplicate docs. The ETL for a datasource will be invoked many times --it must ensure that it only indexes what is necessary.